[hyizhak] iP

hyizhak opened the pull request:
## Bard (iP) Update

> “Your mind is for having ideas, not holding them.” – David Allen

**Bard** (iP) frees your mind from the burden of memorization. It's:

- *Text-based*
- *Easy to learn*  
- ~~Fast~~ **SUPER FAST** to use

Here are some steps to get started:

1. [Download the latest release](https://github.com/hyizhak/ip/releases).
2. Copy the jar file into an empty folder. Open a command window in that folder.
3. Run the command 
```bash
java -jar "bard.jar"
```
3. Add your tasks.
6. Let Bard manage them for you :wink:

### Features

- Managing tasks
- Managing deadlines and events
- *Reminders* (coming soon)

If you are a Java programmer, you can use Bard to practice Java too. Here’s the `main` method for reference:

```java
public class Bard {
    public static void main(String[] args) {
        Bard bard = new Bard();
        bard.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

hyizhak committed:
Accomplish Level 0. Rename, Greet, Exit

hyizhak committed:
Accomplish Level 1. Echo

hyizhak committed:
Accomplish Level 2. Add, List

hyizhak committed:
Accomplish Level 3. Mark as Done

hyizhak committed:
Accomplish Level 4. ToDos, Events, Deadlines

hyizhak committed:
Accomplish Level 5. Handle Errors

hyizhak committed:
Accomplish Level 6. Delete

hyizhak committed:
Accomplish Level 7. Save

hyizhak committed:
Merge branch 'branch-Level-7'
Merge Level-7 Save Branch to the master

hyizhak committed:
Accomplish Level 8. Dates and Times

hyizhak committed:
Merge branch 'branch-Level-8'

TanJieHaoAmos commented on src/main/java/Bard.java:
```diff
@@ -0,0 +1,164 @@
+    public static void main(String[] args){
+        Bard bard = new Bard();
+        bard.run();
+    }
```
No JavaDocs written for the methods and the class

TanJieHaoAmos commented on src/main/java/BardException.java:
```diff
@@ -0,0 +1,10 @@
+public class BardException extends Exception {
```
Similarly, no JavaDocs written for this class.

TanJieHaoAmos replied:
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/Bard.java:
```diff
@@ -0,0 +1,164 @@
+import java.time.LocalDate;
```
Missing package for the class. 
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/BardStorage.java:
```diff
@@ -0,0 +1,83 @@
+import java.io.File;
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/DateParser.java:
```diff
@@ -0,0 +1,68 @@
+import java.time.DayOfWeek;
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/DateParser.java:
```diff
@@ -0,0 +1,68 @@
+import java.time.format.DateTimeParseException;
+
+public class DateParser {
+    static final DateTimeFormatter INPUT_HOUR_FORMAT = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
```
Missing JavaDocs for the class and its methods.

TanJieHaoAmos commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDate;
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDate;
+
+public class Deadline extends Task {
+    protected LocalDate by;
```
Missing JavaDocs for class and its methods

TanJieHaoAmos commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,32 @@
+public class Event extends Task {
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,32 @@
+    public String toFileString() {
+        return "E | " + super.toFileString() + " | " + from + "-" + to;
+    }
+}
```
Missing JavaDocs for the class and its methods.

TanJieHaoAmos commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+public class Task {
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
```
Missing JavaDocs for the class and its methods.

TanJieHaoAmos commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,18 @@
+public class Todo extends Task {
```
Class is not inside a package.
https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Put%20every%20class%20in%20a%20package

TanJieHaoAmos commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,18 @@
+    public String toFileString() {
+        return "T | " + super.toFileString();
+    }
+}
```
Missing JavaDocs for the class and its methods.

TanJieHaoAmos reviewed (commented):
Looks good. Just missing the packages for the classes as well as the javadocs. Code looks clean.

LsmnBmnc commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public String toString() {
```
Good override

Zonnie-23 commented on src/main/java/DateParser.java:
```diff
@@ -0,0 +1,68 @@
+            case "fri", "friday" -> DayOfWeek.FRIDAY;
+            case "sat", "saturday" -> DayOfWeek.SATURDAY;
+            case "sun", "sunday" -> DayOfWeek.SUNDAY;
+            default -> throw new IllegalArgumentException("Invalid day: " + day);
```
Interesting use of switch case to directly return the right enum variable . The formatting looks correct

Zonnie-23 commented on src/main/java/Bard.java:
```diff
@@ -0,0 +1,164 @@
+        tasks.remove(taskNumber - 1);
+        System.out.println(" Now you have " + tasks.size()
+                + " tasks in the list." + "\n" + horizontalLine);
+    }
```
Would be good to package all functions above in a command package/parser package

Zonnie-23 reviewed (commented):
Overall, in terms of coding syntax, it is mostly compliant. The only missing things are the javaDoc and the packaging of the classes. Code is easily understandable, although packaging it into packages would make it easier to understand how the various function interact.

hyizhak committed:
Accomplish A-MoreOOP

hyizhak committed:
Merge branch 'A-MoreOOP'

hyizhak committed:
Accomplish A-Packages

hyizhak committed:
Merge branch 'A-Packages'

hyizhak committed:
Merge branch 'add-gradle-support'

hyizhak committed:
Accomplish A-JUnit

hyizhak committed:
Merge branch 'A-JUnit'

hyizhak committed:
Accomplish A-JavaDoc

hyizhak committed:
Accomplish A-CodingStandard

hyizhak committed:
Accomplish Level 9. Find

hyizhak committed:
Merge branch 'branch-A-JavaDoc'

hyizhak committed:
Merge branch 'branch-A-CodingStandard'

hyizhak committed:
Merge branch 'branch-Level-9'

hyizhak committed:
Accomplish Level 10. GUI

hyizhak committed:
Merge branch 'branch-Level-10'

hyizhak committed:
feat: Add JavaFX dependencies and configure Shadow JAR

- Added JavaFX dependencies for Windows, macOS, and Linux, ensuring cross-platform compatibility.
- Configured the Shadow JAR plugin to package dependencies into a single executable JAR.
- Defined the `shadowJar` task with `archiveBaseName = "bard"` for cleaner artifact naming.
- Set `mainClass` to `bard.Launcher` under the `application` plugin.

These changes streamline the build process, making it easier to package and run the application as a standalone executable.

hyizhak committed:
feat: Add assertions in command parser and main program; set up exit function from Text UI

- Added assertions in the command parser to validate critical assumptions before executing commands.
- Integrated assertions in the main program to ensure that key invariants are maintained.
- Implemented the exit function from the Text UI, allowing for a controlled shutdown of the application.

These changes enhance the overall robustness and reliability of the application, ensuring that failures in critical areas are caught early and the exit behavior is clearly defined.

hyizhak committed:
refactor: Enhance task command and date parsing code quality

- Refactored the command parser's createTask method:
  - Replaced the if-else chain with a switch statement for better clarity.
  - Introduced guard clauses and input trimming to validate command inputs early.
  - Improved error handling by throwing descriptive exceptions for missing or malformed arguments.

- Refactored the DateParser class:
  - Consolidated time parsing logic to avoid redundancy and improve maintainability.
  - Enhanced input validation and error messaging for both date and day formats.
  - Added a private constructor to enforce the utility class pattern.

These improvements result in more robust, readable, and maintainable parsing logic across the application.

hyizhak committed:
refactor: Enhance task command parsing code quality

- Replaced switch statement in CommandParser.parse with enhanced switch

hyizhak committed:
Merge pull request #2 from hyizhak/branch-A-Assertions

feat: Add assertions in command parser and main program; set up exit function from Text UI

hyizhak committed:
chore: Resolve merge conflicts in CommandParser.java for PR

- Merged latest changes from master into the PR branch.
- Manually resolved conflicts in src/main/java/bard/parser/CommandParser.java by integrating modifications from both branches.
- Ensured that command parsing logic remains consistent and all tests pass post-resolution.
- Cleaned up redundant code and aligned with upstream changes to maintain feature integrity.

hyizhak committed:
Merge pull request #3 from hyizhak/branch-A-CodeQuality

Code quality enhancement

hyizhak committed:
ci: set up cross-platform Java CI pipeline with GitHub Actions

- Added a new workflow file ("Java CI") to automatically build and test the project on push and pull_request events.
- Configured a build matrix to run on ubuntu-latest, macos-latest, and windows-latest for broad platform coverage.
- Included steps to:
  - Check out the repository (with a specific checkout for the master branch).
  - Merge the current commit using a forceful checkout.
  - Validate the Gradle Wrapper via the gradle/wrapper-validation-action.
  - Set up JDK 17 with JavaFX support using actions/setup-java.
  - Build the project and run tests using './gradlew check'.
  - Execute IO redirection tests on Linux, macOS, and Windows with respective scripts.

These changes ensure consistent, automated testing and build verification across multiple operating systems.

hyizhak committed:
Merge pull request #4 from hyizhak/branch-A-CI

CI: set up cross-platform Java CI pipeline with GitHub Actions

hyizhak committed:
Correct the directory name to workflows

hyizhak committed:
docs: add missing Javadoc comments and enforce code style guidelines

- Added comprehensive Javadoc comments to public classes and methods across the project (e.g., TaskList, CommandParser, DateParser, etc.).
- Reorganized import orders and adjusted whitespace to address Checkstyle violations.
- Ran Spotless to auto-apply formatting fixes, ensuring that the code conforms to our style rules.

These changes enhance code readability, documentation quality, and maintain consistent style throughout the codebase.

hyizhak committed:
Modify the workflow so it only runs on Linux

hyizhak committed:
Modify the workflow so it only runs on Linux

hyizhak committed:
Merge pull request #5 from hyizhak/branch-A-CI

Add required JavaDoc so the code can pass CI

hyizhak committed:
feat: Add task sorting and integrate into chatbot functionality

- Implemented a new compareTo method for tasks:
  - Incomplete tasks are prioritized over completed ones.
  - Deadline tasks are sorted based on their 'by' time.
  - Event tasks are sorted based on their 'from' time.
- Ensured tasks marked as done always appear at the bottom.
- Integrated the updated sorting logic into the chatbot functionality to enhance task management.

This update improves task prioritization and overall user experience within the chatbot.

hyizhak committed:
Merge pull request #6 from hyizhak/branch-C-Sort

feat: Add task sorting and integrate into chatbot functionality

hyizhak committed:
Add more JUnit testing for Task class

hyizhak committed:
Modify the styling css to highlight errors

hyizhak committed:
Merge pull request #7 from hyizhak/branch-A-MoreTesting

Add more JUnit testing for Task class

hyizhak committed:
Merge pull request #8 from hyizhak/branch-A-BetterGui

Modify the styling css to highlight errors

hyizhak committed:
Adjust the logic of storage-related exception

hyizhak committed:
Merge pull request #9 from hyizhak/branch-A-MoreErrorHandling

Adjust the logic of storage-related exception

hyizhak committed:
Add the user guide in readme.md

hyizhak committed:
Merge pull request #10 from hyizhak/branch-A-UserGuide

Add the user guide in readme.md

hyizhak committed:
Modify the user guide file

hyizhak committed:
Upload UI screenshot

hyizhak committed:
release: version 0.3

hyizhak committed:
fix: remove all sout so error information is handled consistently

hyizhak committed:
fix: update JUnit test to match the new error info

hyizhak committed:
fix: update text ui test

hyizhak committed:
docs: Change the readme
